### ISSUE TITLE
Same parameter with two different types should error

### ISSUE THREAD

**Issue by kevinburke1** [sqlc-dev/sqlc#297]
If I do something like

```sql
CREATE TABLE authors (
  id   BIGSERIAL PRIMARY KEY,
  name text      NOT NULL
);

-- name: UpdateAuthor :exec
UPDATE authors
SET name = $1
WHERE id = $1;
```

In theory this should error because $1 must either be a number or a text. PostgreSQL returns the following error when executing this query:

```
ERROR:  invalid input syntax for type bigint: "foo"
```

In practice sqlc is happy to allow it, generating the following code:

```go
// Code generated by sqlc. DO NOT EDIT.
// source: query.sql

package querytest

import (
	"context"
)

const updateAuthor = `-- name: UpdateAuthor :exec
UPDATE authors
SET name = $1
WHERE id = $1
`

func (q *Queries) UpdateAuthor(ctx context.Context, name string) error {
	_, err := q.db.ExecContext(ctx, updateAuthor, name)
	return err
}
```

**Comment by kevinburkemeter** [comment]
In Postgres, not sure whether MySQL has the same problem.

**Comment by mightyguava** [comment]
oh i think https://github.com/kyleconroy/sqlc/pull/306 is a step towards this, as it no longer drops the second `$1` in the initial parsing step

**Comment by kyleconroy** [comment]
This is fixed in [v1.23.0](https://docs.sqlc.dev/en/latest/reference/changelog.html#v1-23-0) by enabling the [database-backed query analyzer](https://docs.sqlc.dev/en/latest/howto/generate.html#enhanced-analysis-with-managed-databases). We added a [test case](https://github.com/sqlc-dev/sqlc/tree/main/internal/endtoend/testdata/invalid_params_type_mismatch/postgresql/pgx) for this issue so it won’t break in the future.

You can play around with the [working example on the playground](https://play.sqlc.dev/p/cda4ed7dbdac9dda6c7f4f8d52ae1c80a4cd56f4ad297fbaf490d10a3b0f9dc9)
